Mathlib/LinearAlgebra/Matrix/FixedDetMatrices.lean

```diff
@@ -128,110 +130,106 @@ private lemma A_a_ne_zero {A : Δ m} (ha : A.1 1 0 = 0) (hm : m ≠ 0) : A.1 0 0
+  have h1 : 0 < |A.1 1 1| := (abs_nonneg _).trans_lt h11
+  have h2 : 0 < |A.1 0 0| := abs_pos.mpr h00.ne'
   fin_cases i <;> fin_cases j
-  · simpa only [Fin.zero_eta, ← A_c_eq_zero h.1, abs_mul]
-    using (le_mul_iff_one_le_right h2).mpr h1
-  · simpa only [Fin.zero_eta, Fin.mk_one, ← A_c_eq_zero h.1, abs_mul]
-    using h.2.2.2.le.trans (le_mul_of_one_le_left h1.le h2)
-  · simp only [Fin.mk_one, Fin.isValue, Fin.zero_eta, h.1, abs_zero, abs_nonneg]
-  · simpa only [Fin.mk_one, ← A_c_eq_zero h.1, abs_mul] using (le_mul_iff_one_le_left h1).mpr h2
-
-lemma reps_zero_empty : (reps 0) = ∅ := by
-  rw [reps]
-  refine Set.eq_empty_iff_forall_not_mem.mpr fun A h ↦ ?_
-  obtain ⟨h₁, h₂, -, h₄⟩ := Set.mem_setOf.mpr h
-  simp only [eq_zero_of_ne_zero_of_mul_left_eq_zero h₂.ne' <| A_c_eq_zero h₁, abs_zero] at h₄
-  exact ((abs_nonneg _).trans_lt h₄).false
+  · simpa only [← abs_mul, A_c_eq_zero h10] using (le_mul_iff_one_le_right h2).mpr h1
+  · simpa only [← abs_mul, A_c_eq_zero h10] using h11.le.trans (le_mul_of_one_le_left h1.le h2)
+  · simp_all
+  · simpa only [← abs_mul, A_c_eq_zero h10] using (le_mul_iff_one_le_left h1).mpr h2
+
+@[simp]
+lemma reps_zero_empty : reps 0 = ∅ := by
+  rw [reps, Set.eq_empty_iff_forall_not_mem]
+  rintro A ⟨h₁, h₂, -, h₄⟩
+  suffices |A.1 0 1| < 0 by linarith [abs_nonneg (A.1 0 1)]
+  have := A_c_eq_zero h₁
+  simp_all [h₂.ne']
 
 noncomputable instance repsFintype (k : ℤ) : Fintype (reps k) := by
-  by_cases hk : k = 0
-  · rw [hk, reps_zero_empty]
-    exact Set.fintypeEmpty
-  · let H := Finset.Icc (-|k|) |k|
-    let H4 := Fin 2 → Fin 2 → H
-    apply Fintype.ofInjective (β := H4)
-      (f := fun (M : reps k) ↦ fun i j ↦ ⟨M.1.1 i j, reps_entries_le_m' hk M.2 i j⟩)
-    intro M N h
+  let H := Finset.Icc (-|k|) |k|
+  let H4 := Fin 2 → Fin 2 → H
+  apply Fintype.ofInjective (β := H4) (f := fun M i j ↦ ⟨M.1.1 i j, reps_entries_le_m' M.2 i j⟩)
+  intro M N h
+  ext i j
+  simpa only [Subtype.mk.injEq] using congrFun₂ h i j
+
+@[simp]
+lemma S_smul_four (A : Δ m) : S • S • S • S • A = A := by
+  simp only [smul_def, ← mul_assoc, S_mul_S_eq, neg_mul, one_mul, mul_neg, neg_neg, Subtype.coe_eta]
+
+@[simp]
+lemma T_S_rel (A : Δ m) : S • S • S • T • S • T • S • A = T⁻¹ • A := by
+  have : S • S • S • T • S • T • S = T⁻¹ := by
     ext i j
-    simpa only [Subtype.mk.injEq] using congrFun₂ h i j
+    fin_cases i <;> fin_cases j <;> rfl
+  simp_rw [← this, ← smul_assoc]
 
-lemma reduce_mem_reps {m : ℤ} (hm : m ≠ 0) : ∀ A : Δ m, reduce A ∈ reps m := by
-  apply reduce_rec
-  · intro A h
+lemma reduce_mem_reps {m : ℤ} (hm : m ≠ 0) (A : Δ m) : reduce A ∈ reps m := by
+  induction A using reduce_rec with
+  | step A h1 h2 => simpa only [reduce_reduceStep h1] using h2
+  | base A h =>
     have hd := A_d_ne_zero h hm
     by_cases h1 : 0 < A.1 0 0
-    focus simp only [reduce_of_pos h h1]
-    swap
-    focus simp only [reduce_of_not_pos h h1]
-    swap
-    all_goals
+    · simp only [reduce_of_pos h h1]
```

Comment:
Here I expanded the proof quite a bit. But I think that the massive `all_goals simp only` is quite impenetrable. And hard to maintain if it might break at some point in the future.

Reply:
(I'm responsible for that, I guess; I tried to combine the two massive `simp only`s that used to be there before, as they were fairly similar.)